### ISSUE TITLE
scala.MatchError on reading metadata

### ISSUE THREAD

**Issue by drudim** [elastic/elasticsearch-hadoop#924]
### What kind an issue is this?

- [x] Bug report.

### Issue description

Getting `scala.MatchError` while trying to read `_metadata`.

### Steps to reproduce

Code:
```
# Fails
scala> spark.read.format("org.elasticsearch.spark.sql").option("es.read.metadata", "true").option("es.nodes", "xxx").load("yyy/zzz").select("_metadata")

# Fails 
scala> spark.read.format("org.elasticsearch.spark.sql").option("es.read.metadata", "true").option("es.nodes", "xxx").load("yyy/zzz").select("_metadata", "some_field")

# Works!
scala> spark.read.format("org.elasticsearch.spark.sql").option("es.read.metadata", "true").option("es.nodes", "xxx").load("yyy/zzz").select("some_field", "_metadata")
```

Strack trace:
```
17/01/30 21:54:51 ERROR TaskSetManager: Task 0 in stage 2.0 failed 1 times; aborting job
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 2.0 failed 1 times, most recent failure: Lost task 0.0 in stage 2.0 (TID 2, localhost): scala.MatchError: true (of class java.lang.Boolean)
       	at org.apache.spark.sql.catalyst.CatalystTypeConverters$MapConverter.toCatalystImpl(CatalystTypeConverters.scala:202)
       	at org.apache.spark.sql.catalyst.CatalystTypeConverters$MapConverter.toCatalystImpl(CatalystTypeConverters.scala:194)
       	at org.apache.spark.sql.catalyst.CatalystTypeConverters$CatalystTypeConverter.toCatalyst(CatalystTypeConverters.scala:103)
       	at org.apache.spark.sql.catalyst.CatalystTypeConverters$$anonfun$createToCatalystConverter$2.apply(CatalystTypeConverters.scala:403)
       	at org.apache.spark.sql.execution.RDDConversions$$anonfun$rowToRowRdd$1$$anonfun$apply$2.apply(ExistingRDD.scala:67)
       	at org.apache.spark.sql.execution.RDDConversions$$anonfun$rowToRowRdd$1$$anonfun$apply$2.apply(ExistingRDD.scala:64)
       	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
       	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
       	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
       	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:370)
       	at org.apache.spark.sql.execution.SparkPlan$$anonfun$4.apply(SparkPlan.scala:246)
       	at org.apache.spark.sql.execution.SparkPlan$$anonfun$4.apply(SparkPlan.scala:240)
       	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:803)
       	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:803)
       	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
       	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:319)
       	at org.apache.spark.rdd.RDD.iterator(RDD.scala:283)
       	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:70)
       	at org.apache.spark.scheduler.Task.run(Task.scala:86)
       	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:274)
       	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
       	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
       	at java.lang.Thread.run(Thread.java:745)
``` 

### Version Info

OS: Amazon Linux AMI release 2016.09
JVM: OpenJDK 64-Bit Server VM, Java 1.8.0_111, Scala 2.11.8
Hadoop/Spark: 2.7 / 2.0.2
ES-Hadoop: 5.1.1
ES: 2.4.1

**Comment by rawata** [comment]
similar things was reported here
https://discuss.elastic.co/t/elasticsearch-dependency-on-zeppelin-and-emr-breaks-the-ability-to-read-parquet-files/66818

**Comment by drudim** [comment]
Managed to fix it with a hack here https://github.com/drudim/elasticsearch-hadoop/commit/ece21fcec0368cd75d4d3fe70cf0385aa6568065 , but it isn't a solution that is suitable for a PR to the main repo. So, I am still digging to the code and thinking about decoupling of `INTERNAL_ES_TARGET_FIELDS` and fields provided by Spark SQL in `LogicalRelation`. As I understood it fails because of the different order of fields in `r` (row) and in `outputTypes` here https://github.com/apache/spark/blob/a6155135690433988aa0cbf22f260f52a235e9f5/sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala#L60

It looks like the order of fields in a row is coupled to `INTERNAL_ES_TARGET_FIELDS` via `getScrollFields`
1) https://github.com/elastic/elasticsearch-hadoop/blob/e80ad2e05889bdfabbddd835f31a3a0fea086c1c/spark/sql-13/src/main/scala/org/elasticsearch/spark/sql/SchemaUtils.scala#L278
2) https://github.com/elastic/elasticsearch-hadoop/blob/b8784b4002d3f270b36cb834680cbe3aa0a23574/mr/src/main/java/org/elasticsearch/hadoop/cfg/Settings.java#L152

Not sure about these assumptions though :)

**Comment by jbaiera** [comment]
@drudim  Thanks for the open pull requests. I'll try and find some time to review them soon!

**Comment by drudim** [comment]
@jbaiera any updates on this one?

**Comment by jbaiera** [comment]
This was fixed with #930 and included in 5.4.1 and 6.0.0. Closing this issue as it was left open.
